File: scripts/setup/sync-assets.js
#!/usr/bin/env node

/**
 * Sync Shared Assets
 * Synchronizes common assets (styles, icons, mocks) across all framework apps
 */

const fs = require('fs');
const path = require('path');

const ASSETS_DIR = path.join(__dirname, '..', '..', 'assets');
const APPS_DIR = path.join(__dirname, '..', '..', 'apps');
const FRAMEWORKS_CONFIG = path.join(__dirname, '..', '..', 'frameworks.json');

/**
 * Get file/directory size recursively
 */
function getDirectorySize(dirPath) {
  let totalSize = 0;
  
  if (!fs.existsSync(dirPath)) return 0;
  
  function calculateSize(itemPath) {
    const stats = fs.statSync(itemPath);
    if (stats.isDirectory()) {
      fs.readdirSync(itemPath).forEach(item => {
        calculateSize(path.join(itemPath, item));
      });
    } else {
      totalSize += stats.size;
    }
  }
  
  calculateSize(dirPath);
  return totalSize;
}

/**
 * Format bytes to human readable format
 */
function formatBytes(bytes) {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + ' ' + sizes[i];
}

/**
 * Copy directory with file count tracking
 */
function copyDirectoryWithStats(source, target) {
  if (!fs.existsSync(source)) return { files: 0, size: 0 };
  
  const stats = { files: 0, size: 0 };
  
  function copyRecursive(src, dest) {
    const sourceStats = fs.statSync(src);
    
    if (sourceStats.isDirectory()) {
      fs.mkdirSync(dest, { recursive: true });
      fs.readdirSync(src).forEach(item => {
        copyRecursive(path.join(src, item), path.join(dest, item));
      });
    } else {
      fs.copyFileSync(src, dest);
      stats.files++;
      stats.size += sourceStats.size;
    }
  }
  
  copyRecursive(source, target);
  return stats;
}

/**
 * Get framework configuration
 */
function getFrameworkConfig(appName) {
  try {
    const config = JSON.parse(fs.readFileSync(FRAMEWORKS_CONFIG, 'utf8'));
    const framework = config.frameworks.find(f => f.id === appName || f.dir === appName);
    return framework || { assetsDir: 'public' }; // Default to 'public' if not found
  } catch (error) {
    console.warn(`Warning: Could not load frameworks.json, using defaults for ${appName}`);
    return { assetsDir: 'public' };
  }
}

/**
 * Sync assets for a specific app
 */
function syncAppAssets(appName, appPath) {
  const frameworkConfig = getFrameworkConfig(appName);
  const assetsDir = path.join(appPath, frameworkConfig.assetsDir || 'public');
  
  // Ensure assets directory exists
  fs.mkdirSync(assetsDir, { recursive: true });
  
  // Copy entire assets directory
  const stats = copyDirectoryWithStats(ASSETS_DIR, assetsDir);
  
  return [{
    type: 'all assets',
    files: stats.files,
    size: stats.size
  }];
}

/**
 * Main sync function
 */
async function syncAssets() {
  const startTime = Date.now();
  
  try {
    console.log('🔄 Synchronizing shared assets across framework apps...\n');
    
    // Check if directories exist
    if (!fs.existsSync(APPS_DIR)) {
      throw new Error('Apps directory not found');
    }
    
    if (!fs.existsSync(ASSETS_DIR)) {
      throw new Error('Assets directory not found');
    }
    
    // Get all app directories
    const apps = fs.readdirSync(APPS_DIR).filter(dir => {
      const appPath = path.join(APPS_DIR, dir);
      return fs.statSync(appPath).isDirectory() && dir !== 'node_modules';
    });
    
    if (apps.length === 0) {
      throw new Error('No app directories found');
    }
    
    const allResults = [];
    
    // Sync assets for each app
    for (const app of apps) {
      const appPath = path.join(APPS_DIR, app);
      console.log(`📁 ${app}:`);
      
      const syncResults = syncAppAssets(app, appPath);
      
      let totalFiles = 0;
      let totalSize = 0;
      
      syncResults.forEach(result => {
        if (result.files > 0) {
          console.log(`   ${result.type}: ${result.files} files (${formatBytes(result.size)})`);
          totalFiles += result.files;
          totalSize += result.size;
        }
      });
      
      console.log(`   ✅ Total: ${totalFiles} files (${formatBytes(totalSize)})\n`);
      
      allResults.push({
        app,
        files: totalFiles,
        size: totalSize,
        syncResults
      });
    }
    
    const duration = Date.now() - startTime;
    const grandTotalFiles = allResults.reduce((sum, result) => sum + result.files, 0);
    const grandTotalSize = allResults.reduce((sum, result) => sum + result.size, 0);
    
    // Summary
    console.log('✅ Asset Synchronization Complete');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log(`📋 Apps processed: ${apps.length}`);
    console.log(`📄 Total files synced: ${grandTotalFiles}`);
    console.log(`💾 Total data synced: ${formatBytes(grandTotalSize)}`);
    console.log(`⏱️  Completed in: ${duration}ms`);
    console.log(`\n🎯 Apps synced: ${apps.join(', ')}\n`);
    
  } catch (error) {
    console.error('❌ Error syncing assets:', error.message);
    process.exit(1);
  }
}

if (require.main === module) {
  syncAssets();
}

module.exports = { 
  syncAssets, 
  syncAppAssets, 
  formatBytes, 
  getDirectorySize 
};

File: scripts/setup/sync_assets.py
"""
Sync assets across all framework apps.
All shared assets (styles, images, fonts, etc) are stored in ./assets
This script then copies them to each framework's source.
This is usually within ./apps/[app-name]/public
but can be customized with the `assetsDir` config option in frameworks.json
"""

import shutil
from pathlib import Path
from typing import Dict, Any

import click
from rich.progress import Progress, SpinnerColumn, TextColumn

import sys
from pathlib import Path
import time
sys.path.append(str(Path(__file__).parent.parent))

from common import (
    console, get_project_root, get_frameworks_config,
    show_header, show_success, show_error, get_framework_asset_dir
)


def sync_app_assets(app_name: str, app_path: Path, assets_dir: Path, frameworks_config: Dict[str, Any]) -> bool:
    """Sync assets for a single app."""
    try:
        # Get framework-specific asset directory
        asset_dir_name = get_framework_asset_dir(app_name, frameworks_config)
        app_asset_path = app_path / asset_dir_name
        
        # Create asset directory if it doesn't exist
        app_asset_path.mkdir(exist_ok=True)
        
        # Copy all assets
        for item in assets_dir.iterdir():
            if item.is_file():
                target = app_asset_path / item.name
                shutil.copy2(item, target)
            elif item.is_dir():
                target = app_asset_path / item.name
                if target.exists():
                    shutil.rmtree(target)
                shutil.copytree(item, target)
        
        return True
    except Exception as e:
        show_error(f"Failed to sync assets for {app_name}: {e}")
        return False


@click.command()
def sync_assets():
    """Sync assets from the global assets directory to all framework apps."""
    show_header("Asset Sync", "Copying shared assets to all framework applications")
    
    project_root = get_project_root()
    assets_dir = project_root / "assets"
    apps_dir = project_root / "apps"
    frameworks_config = get_frameworks_config()
    
    if not assets_dir.exists():
        show_error(f"Assets directory not found: {assets_dir}")
        return
    
    frameworks = [fw["id"] for fw in frameworks_config.get("frameworks", [])]
    failed_apps = []
    
    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        console=console,
    ) as progress:
        tasks = {}
        for framework in frameworks:
            task = progress.add_task(f"🔄️ Syncing {framework}...", total=1)
            tasks[framework] = task
            time.sleep(0.25)
            app_path = apps_dir / framework
            if not app_path.exists():
                show_error(f"App directory not found: {app_path}")
                failed_apps.append(framework)
                progress.update(task, description=f"❌ {framework} (failed)", completed=1)
                continue

            success = sync_app_assets(framework, app_path, assets_dir, frameworks_config)
            if not success:
                failed_apps.append(framework)
                progress.update(task, description=f"❌ {framework} (failed)", completed=1)
            else:
                progress.update(task, description=f"✔️ Synced {framework}", completed=1)
    
    # Summary
    success_count = len(frameworks) - len(failed_apps)
    show_success(f"Successfully synced assets for {success_count}/{len(frameworks)} apps")
    
    if failed_apps:
        show_error(f"Failed to sync: {', '.join(failed_apps)}")


if __name__ == "__main__":
    sync_assets()
